### PR TITLE
feat: add LOD mutation APIs (distance/profile)

### PR DESCRIPTION
**Pull request by 0x74h51N**
This PR enables updating LOD distances and hysteresis at runtime or remove.

updateLOD presets can be tweaked on the fly (UI/settings) safely.
Preserves invariants: squared-distance storage, strictly increasing thresholds.
Avoids direct writes to levels[i].

### Adds:
- updateLOD(levelIndex: number, distance?: number, hysteresis?: number): this;
- updateShadowLOD(levelIndex: number, distance?: number, hysteresis?: number): this;
- updateAllLOD(distances?: number[], hysteresis?: number | number[]): this;
- updateAllShadowLOD(distances?: number[], hysteresis?: number | number[]): this;
- removeLOD(levelIndex: number, removeObject?: boolean): this;
- Internal helpers: updateLevel(...), updateAllLevels(...), disposeLOD(...)

Notes:
- Error handling: methods throw on invalid lists, OOB indices, NaN distances, or non-increasing sequences. If the target LOD list doesn’t exist => no-op.
- updateAllLevels works with optional distance or hysteresis. If distances are provided, validates order. If first distance is 0, handle gracefully. Throws if order is wrong.
- removeLOD: removes whole list if only LOD0 remains, mirrors remove on shadow list, and by default removes/disposes the LOD object (geo + mat).
- disposeLOD helper is added to safely release geometry/material resources when a LOD is removed, to avoid memory leaks.

### Discussion:
- Is updating all nodes with optional distance really needed? Feels like an edge case, adds more control logic. Imho without distance, updating LOD isn’t that important, but on the other hand it can be useful for dev GUIs to test hysteresis or distances separately.
- Edge case: removing all LODs could be useful, e.g. when user wants to change geometry segments at runtime for quality settings. Needs removing all LODs, and looping with indices can get tricky.
- Should we check if an object is still used in another LOD before disposing (stillUsed check)? Might be overkill, but safer.
- Shadow LOD removal has bug potential. Right now addShadowLOD works with independent distance input, separate from normal LODs. Internally addShadowLOD just checks if the geometry already exists, if so, it reuses the same object, sets object.castShadow = true, and doesn’t create a new mesh or apply any different shadow process. That means there’s no actual “lower quality” shadow rendering per distance. In remove phase this can break, because shadow LODs are removed by index: wrong shadow LOD could be removed or unexpected shadows may still show.
A better approach might be to remove the separate shadow LOD structure entirely and instead manage castShadow / receiveShadow options directly on objects when adding them to LOD (similar to how Three.js handles it). That would also solve the **current receiveShadow bug** and remove the need for updateShadowLOD. But this would require changes to frustum culling / render order, so not for this PR, still, it’s enough reason not to add a dedicated removeShadowLOD.

### CODE REVIEW

**Comment by agargaro** [comment]
Nice work! I've refactored a bit the code :) I need to review the `removeLOD` method, I'll do it tomorrow.

We should add also the `removeShadowLOD` method I guess.

related #29

File: src/core/feature/LOD.ts [hunk 1]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
+    level.distance = distance ** 2; // to avoid to use Math.sqrt every time
+  }
+
+  if (hysteresis != null) {
+    level.hysteresis = hysteresis;
   }
-  if (hysteresis != null) entry.hysteresis = hysteresis;
+
+  if (sort) {
+    renderList.levels.sort((a, b) => a.distance - b.distance);
```
Comment (0x74h51N) [inline src/core/feature/LOD.ts:244]: Why sort? 
This method doesn’t change geometry, it only changes distance and hysteresis. If the user gives a bigger number to the next LOD level, we can throw an error; sorting will break the whole LOD level geometry ordering.

File: src/core/feature/LOD.ts [hunk 2]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
-  for (let i = 0; i < n; i++) {
-    const d = distances[i];
-    if (d == null || Number.isNaN(d)) {
-      throw new Error(`LOD distance at index ${i} is invalid (${d}).`);
-    }
-    if (i > 0 && d <= prev) {
-      throw new Error(`LOD distances must be strictly increasing: d[${i - 1}]=${prev} < d[${i}]=${d}`);
-    }
-    const h = Array.isArray(hysteresis) ? hysteresis[i] : hysteresis;
-    this.changeLevel(renderList, start + i, d, h);
```
Comment (0x74h51N) [inline src/core/feature/LOD.ts]: This LOD distance check provides a clear error message, but it should run before any changes are applied. I mistakenly refactored it here.

**Comment by 0x74h51N** [comment]
> Nice work! I've refactored a bit the code :) I need to review the `removeLOD` method, I'll do it tomorrow.
> 
> We should add also the `removeShadowLOD` method I guess.
> 
> related #29

Thanks! I’ve left some comments.
I think a few of these changes may break the LOD level order and introduce runtime issues. It’d be better to discuss before changing behavior like this.

About removeShadowLOD, I agree it’s useful, users might want to delete only a shadow level. Note that when a render LOD is removed, the corresponding shadow level is already removed as well.

File: src/core/feature/LOD.ts (hunk 1 above)
Comment (0x74h51N) [inline src/core/feature/LOD.ts:244]: Why sort? Sorting levels after changing distance or hysteresis will break the geometry order. The current design allowed adjusting levels of detail at runtime without altering the geometry order, for example, using more optimized distances for mobile devices or higher quality for high-end systems.

File: src/core/feature/LOD.ts [hunk 3]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
-    prev = d;
+  for (let i = 0; i < count; i++) {
+    const distance = distances?.[i];
+    const _hysteresis = Array.isArray(hysteresis) ? hysteresis[i] : hysteresis;
+    this.updateLevel(renderList, i, distance, _hysteresis, false);
   }
+
+  renderList.levels.sort((a, b) => a.distance - b.distance);
```
Comment (0x74h51N) [inline src/core/feature/LOD.ts:269]: This sorting will break the geometry order as well. It’s not needed. Validating distances before applying changes would be best.

File: src/core/feature/LOD.ts [hunk 4]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
-    if (d == null || Number.isNaN(d)) {
-      throw new Error(`LOD distance at index ${i} is invalid (${d}).`);
-    }
-    if (i > 0 && d <= prev) {
-      throw new Error(`LOD distances must be strictly increasing: d[${i - 1}]=${prev} < d[${i}]=${d}`);
-    }
-    const h = Array.isArray(hysteresis) ? hysteresis[i] : hysteresis;
-    this.changeLevel(renderList, start + i, d, h);
-    prev = d;
+  for (let i = 0; i < count; i++) {
```
[the remark this answers is not in the document]
Comment (agargaro) [reply src/core/feature/LOD.ts:263]: The `setFirstLODDistance` method is used to set the distance of the main mesh (before adding other LODs).

This is useful for lazy loading high poly geometries, for faster loading times.

File: src/core/feature/LOD.ts (hunk 1 above)
Comment (agargaro) [reply src/core/feature/LOD.ts:244]: I agree that it may seem unnecessary to change the order of geometries/materials, but there are rare cases where it can be useful (especially with materials). I would like to leave this freedom to the user.

File: src/core/feature/LOD.ts [hunk 5]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
+InstancedMesh2.prototype.updateLevel = function (renderList, levelIndex, distance, hysteresis, sort = true) {
+  if (!renderList) throw new Error('Render list is invalid.');
 
-  const entry = levels[levelIndex];
+  const level = renderList.levels[levelIndex];
+  if (!level) throw new Error('Cannot update an empty LOD.');
 
   if (distance != null) {
-    const d2 = distance ** 2;
-    if (Number.isNaN(d2)) throw new Error('Distance is NaN');
```
[the remark this answers is not in the document]
Comment (agargaro) [reply src/core/feature/LOD.ts]: I usually prefer to avoid these checks on primitive types. Consider that even `“2” ** 2` is 4.

File: src/core/feature/LOD.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (0x74h51N) [reply src/core/feature/LOD.ts:263]: This for loop doesn’t use addLOD or addLevel; it uses updateLevel, and setFirstLODDistance doesn’t work in this situation. You’re not keeping the 0-indexed LOD distance at 0, so this ends up changing the 0-indexed LOD’s distance.

File: src/core/feature/LOD.ts [hunk 6]
```diff
@@ -250,99 +226,57 @@ InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometr
...
+
   return this;
 };
 
-InstancedMesh2.prototype.setLODDistance = function (
-  levelIndex,
-  distance,
-  hysteresis
-) {
-  if (levelIndex === 0) return this;
```
Comment (0x74h51N) [inline src/core/feature/LOD.ts]: This was for keeping the 0-indexed LOD distance at 0, why did you remove it?

File: src/core/feature/LOD.ts [hunk 7]
```diff
@@ -90,68 +89,45 @@ declare module '../InstancedMesh2.js' {
...
-     * - No-op when render LODs are not present.
-     * @param distance The distance for this LOD level.
-     * @param hysteresis The hysteresis value for this LOD level.
+     * Updates the LOD settings for all levels.
+     * @param distances The array of distances for each LOD level.
+     * @param hysteresis The hysteresis value(s) for each LOD level.
      * @returns The current `InstancedMesh2` instance.
      */
-    setLODProfile(distances: number[], hysteresis?: number | number[]): this;
+    updateAllLOD(distances: number[] | null, hysteresis?: number | number[]): this;
```
Comment (0x74h51N) [inline src/core/feature/LOD.ts:113]: Why null? undefined would be enough, why allow null?

File: src/core/feature/LOD.ts (hunk 1 above)
Comment (agargaro) [reply src/core/feature/LOD.ts:244]: In that case you should use [0, 60, 80], and I know that it seems user-unfriendly.

But I think that updating also the first level can be useuful with the lazy loading, I need to test it before to be sure.

File: src/core/feature/LOD.ts (hunk 7 above)
Comment (agargaro) [reply src/core/feature/LOD.ts:113]: but it allows you to write `updateAllLOD()` without parameters in typescript.

```ts
return this.updateAllLevels(this.LODinfo?.render, distances, hysteresis);
```

`this.LODinfo?.render` can be null or undefined, it will require a bit more of refactoring, but I'll remove it now

File: src/core/feature/LOD.ts (hunk 1 above)
Comment (0x74h51N) [reply src/core/feature/LOD.ts:244]: The user doesn’t know they must keep the first distance at 0. The first LOD distance is automatically set to 0, but there’s no error handling if the user tries to assign another value. For example, if they pass [60, 80], it will result in an invalid ordering and the geometry will break silently. Sorting here doesn’t make sense, and it’s unnecessary overhead since array sorting isn’t free either.

Comment (agargaro) [reply src/core/feature/LOD.ts]: we can't talk about overhead if we're sorting an array of 3-4 elements.

```ts
instancedMeshLOD.setFirstLODDistance(1000);

setTimeout(() => instancedMeshLOD.addLOD(geoLow, new MeshLambertMaterial({ color: 'orange' }), 500), 1000);
setTimeout(() => instancedMeshLOD.addLOD(geoMid, new MeshLambertMaterial({ color: 'yellow' }), 50), 2000);
setTimeout(() => instancedMeshLOD.addLOD(geoHigh, new MeshLambertMaterial({ color: 'green' }), 0), 3000);
```

Btw this works well, so we can restore validation and avoid to shuffle levels 😄 
The only issue is if the user, set distance = 0, when add the first or the second LOD, but I think it's okay, following the example.

Comment (0x74h51N) [reply src/core/feature/LOD.ts]: Tbh, I don’t understand the example. We’re all on the same page about addLOD or setFirstLODDistance (even if they’re unnecessary and risky to expose publicly, or could be improved). addLOD works fine.
```typescript
InstancedMesh2.prototype.addLOD = function (geometry: BufferGeometry, material: Material | Material[], distance = 0, hysteresis = 0): InstancedMesh2 {
  if (this._parentLOD) {
    throw new Error('Cannot create LOD for this InstancedMesh2.');
  }

  if (!this.LODinfo?.render && distance === 0) {
    throw new Error('Cannot set distance to 0 for the first LOD. Call "setFirstLODDistance" method before use "addLOD".');
  }

  this.setFirstLODDistance(0);

  this.addLevel(this.LODinfo.render, geometry, material, distance, hysteresis);

  return this;
};

...

InstancedMesh2.prototype.addLevel = function (renderList: LODRenderList, geometry: BufferGeometry, material: Material, distance: number, hysteresis: number): InstancedMesh2 {
  const objectsList = this.LODinfo.objects;
  const levels = renderList.levels;
  let index;
  let object: InstancedMesh2;
  distance = distance ** 2; // to avoid to use Math.sqrt every time
...

  for (index = 0; index < levels.length; index++) {
    if (distance < levels[index].distance) break;
  }
//This approach keep to LODs order
  levels.splice(index, 0, { distance, hysteresis, object });
  renderList.count.push(0);

  return object;
};
```

The updating part was meant for runtime. Even if the user calls updateLODs at compile time, TypeScript would catch the errors, but at runtime everything can change. Don’t make me write a Jira-style story, please 😅 Here’s the live example https://github.com/0x74h51N/tarnished-house/pull/1 You can change LOD settings with options

**Comment by 0x74h51N** [comment]
> There are a few things to change in the `removeLOD` method. This should not change the shadow LODs; there will be a separate method for that. I also think we should always remove InstancedMesh2, perhaps by having it returned by the method itself, so that we can dispose of it if necessary (the geometry and material if not shared).
> 
> What do you think?
> 
> For the removeLOD methods, I opened my own branch based on yours, as it's better to treat them separately. Could you remove the method from this PR so we can continue in the other one I'll open?
> 
> After that, I can merge it.

RemoveLOD already returns the IM2. LOD and IM2 are completely different things, removing the LOD at runtime may be for updating the geometry or else be necessary. Even if there is no LOD, keeping the IM2 in the scene and letting the user remove and dispose it is more appropriate. Since Frustum doesn’t work with shared materials, sharing is not really possible, but still in the explanation I had already left a “removeObject” option to the user in case the geometry or mat is shared…

Also, I updated the PR description where I wrote that there’s no real meaning in separating shadowLOD, that the core architecture is problematic, and that possible issues may arise in the remove part as well, did you read it? Moreover, since there’s not much meaning in having a shadowLOD when the LOD itself doesn’t exist, I removed the shadow of the related LOD. If necessary, I can revert this part.

Opening a new branch based on my work, or changing the entire work logic under the name of “refactoring” without discussion, goes against the philosophy of contribution, it’s like taking over someone else’s effort. If you had said from the start that you wanted it separate, I would have opened it separately, normally I only added it here because you asked…

**Comment by agargaro** [comment]
There are projects where only shadowLOD has been set up, where simplified meshes are used only on shadows.
The shadowLOD does not necessarily have the original mesh in the render list, because of this.
So I would prefer to keep the logic separate.

> Opening a new branch based on my work, or changing the entire work logic under the name of “refactoring” without discussion, goes against the philosophy of contribution, it’s like taking over someone else’s effort. If you had said from the start that you wanted it separate, I would have opened it separately, normally I only added it here because you asked…

I understand your concern, and you’re right — I shouldn’t have asked to include this part in the same PR.
I’m currently very busy, and with all the discussions this PR is becoming quite complex. I think it would be better to isolate the `removeLOD` logic so that the main LOD mutation changes remain focused and easier to review.
I’m making some adjustments to your implementation mainly to keep it consistent with the style of the library, which should save time for both of us. The core of your work will remain, and I really appreciate the effort and thought you’ve put into it.
I’ll take care of this part in a separate branch and keep your credit for the original logic, so we can move forward faster without losing your contribution.

**Comment by 0x74h51N** [comment]
> There are projects where only shadowLOD has been set up, where simplified meshes are used only on shadows. The shadowLOD does not necessarily have the original mesh in the render list, because of this. So I would prefer to keep the logic separate.
> 
> > Opening a new branch based on my work, or changing the entire work logic under the name of “refactoring” without discussion, goes against the philosophy of contribution, it’s like taking over someone else’s effort. If you had said from the start that you wanted it separate, I would have opened it separately, normally I only added it here because you asked…
> 
> I understand your concern, and you’re right — I shouldn’t have asked to include this part in the same PR. I’m currently very busy, and with all the discussions this PR is becoming quite complex. I think it would be better to isolate the `removeLOD` logic so that the main LOD mutation changes remain focused and easier to review. I’m making some adjustments to your implementation mainly to keep it consistent with the style of the library, which should save time for both of us. The core of your work will remain, and I really appreciate the effort and thought you’ve put into it. I’ll take care of this part in a separate branch and keep your credit for the original logic, so we can move forward faster without losing your contribution.

I don’t understand what you expect from me. Do you want a new branch specifically for removeLOD? If so, I can open one, but are you asking me to remove what I’ve already done? Which works great? 
If you still want changes, if you really feel the need to interfere, do them after this PR is merged. What exactly are we discussing here? I already wrote a lot of points to discuss in the PR for further more improvements, so what are we discussing right now?
